File: fe/user_management/lib/features/list_users/data/datasources/user_local_data_source.dart
import 'package:objectbox/objectbox.dart';
import 'package:user_management/core/error/exception.dart';
import 'package:user_management/features/list_users/data/models/user_list_model.dart';

import '../../../../core/util/user_mapper.dart';
import '../models/user_model.dart';
import '../objectbox/entities/user.dart';

abstract class UserLocalDataSource {
  ///
  /// Gets the last cached set of users
  /// Throws [CacheException] if no users stored
  Future<UserListModel> getUsers();

  ///
  /// Caches a set of users
  ///
  Future<void> cacheUsers(UserListModel users);
}

class UserLocalDataSourceObjectBoxImpl extends UserLocalDataSource {
  final Box<User> box;
  final IUserMapper mapper;

  UserLocalDataSourceObjectBoxImpl({required this.box, required this.mapper});

  @override
  Future<void> cacheUsers(UserListModel users) async {
    List<User> storedUsers = users.userModels
        .map((userModel) => mapper.mapUserModelToUser(userModel))
        .toList();
    box.putMany(storedUsers);
    return Future.value();
  }

  @override
  Future<UserListModel> getUsers() async {
    List<User> users = box.getAll();
    List<UserModel> userModels = users
        .map((user) => mapper.mapUserToUserModel(user))
        .toList();

    if (userModels.isNotEmpty) {
      return UserListModel(userModels: userModels);
    } else {
      throw CacheException();
    }
  }
}

File: fe/kotlin/UserManagement/app/src/main/java/com/example/usermanagement/features/get_users/data/datasources/UserLocalDataSource.kt
package com.example.usermanagement.features.get_users.data.datasources

import androidx.lifecycle.MutableLiveData
import com.example.usermanagement.features.get_users.data.models.UserModel
import io.reactivex.rxjava3.core.Observable

abstract class UserLocalDataSource {
    abstract fun getUsers(): MutableLiveData<List<UserModel>>
}

class UserLocalDataSourceImpl: UserLocalDataSource() {
    override fun getUsers(): MutableLiveData<List<UserModel>> {
        TODO("Not yet implemented")
    }
}